src/QsCompiler/Transformations/LiftLambdas.cs

```diff
@@ -0,0 +1,243 @@
+                            var localVar = new LocalVariableDeclaration<QsLocalSymbol>(
+                                QsLocalSymbol.NewValidName(sym.Item),
+                                paramType,
+                                new InferredExpressionInformation(false, false),
+                                QsNullable<Position>.Null,
+                                paramName.Range.IsNull
+                                    ? DataTypes.Range.Zero
+                                    : paramName.Range.Item);
+                            return ParameterTuple.NewQsTupleItem(localVar);
+                        }
+                        else if (paramName.Symbol is QsSymbolKind<QsSymbol>.SymbolTuple tup)
+                        {
+                            if (tup.Item.Length == 0 && paramType.Resolution.IsUnitType)
+                            {
+                                // Need an artificial Unit parameter here
+                                var localVar = new LocalVariableDeclaration<QsLocalSymbol>(
+                                    QsLocalSymbol.NewValidName("__lambdaUnitParam__"),
+                                    paramType,
+                                    new InferredExpressionInformation(false, false),
+                                    QsNullable<Position>.Null,
+                                    paramName.Range.IsNull
+                                        ? DataTypes.Range.Zero
+                                        : paramName.Range.Item);
+                                return ParameterTuple.NewQsTupleItem(localVar);
+                            }
+                            else if (tup.Item.Length == 1)
+                            {
+                                return MatchNameWithType(paramType, tup.Item.First());
+                            }
+                            else if (paramType.Resolution is ResolvedTypeKind.TupleType tupType)
+                            {
+                                var subSymbols = tup.Item;
+                                var subSybmolTypes = tupType.Item;
+
+                                if (subSymbols.Length != subSybmolTypes.Length)
+                                {
+                                    throw new ArgumentException("Lambda parameter type length mismatch");
+                                }
+
+                                return ParameterTuple.NewQsTuple(subSymbols
+                                    .Select((symbol, i) => MatchNameWithType(subSybmolTypes[i], symbol))
+                                    .ToImmutableArray());
+                            }
+                            else
+                            {
+                                throw new ArgumentException("Lambda parameter type mismatch");
+                            }
+                        }
+                        else
+                        {
+                            throw new ArgumentException("Lambda parameter type mismatch");
```

Comment:
Callables in Q# can't be defined with missing parameters, so lambdas shouldn't be defined with missing parameters either.

Reply:
You can generate names for the parameters when lifting to work around that. Lambda parameters are designed to be like `let` bindings (no type annotations, and support for discards), rather than global callable declarations. It's convenient to be able to discard parameters since brevity is more important for lambdas than global callable declarations.

Reply:
(Although it might be worth checking if having a discard as a callable parameter actually causes issues. I think it's prevented by the parser, but I'm not sure if any later step currently depends on that.)

Reply:
> Lambda parameters are designed to be like `let` bindings (no type annotations, and support for discards), rather than global callable declarations. It's convenient to be able to discard parameters since brevity is more important for lambdas than global callable declarations.

Okay, thanks for the clarity on the difference. That makes sense.

> You can generate names for the parameters when lifting to work around that. (Although it might be worth checking if having a discard as a callable parameter actually causes issues. I think it's prevented by the parser, but I'm not sure if any later step currently depends on that.)

My preference is that, if it is not OK with the parser, we shouldn't lean into it, even if nothing depends on it. Ideally, we would be able to take the code generated at the end of the transformation steps and be able to feed it back into the compiler from the start, although there are a few caveats for that like the use of the reserved double-underscore identifiers. I will look into generating parameters for these missing parameters, probably using the double-underscore identifiers to avoid colliding with existing symbols.

Reply:
This should be supported now.